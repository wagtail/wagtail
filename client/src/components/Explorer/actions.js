import * as admin from '../../api/admin';
import { createAction } from '../../utils/actions';
import { MAX_EXPLORER_PAGES } from '../../config/wagtailConfig';

const getPageStart = createAction('GET_PAGE_START');
const getPageSuccess = createAction('GET_PAGE_SUCCESS', (id, data) => ({ id, data }));
const getPageFailure = createAction('GET_PAGE_FAILURE', (id, error) => ({ id, error }));

/**
 * Gets a page from the API.
 */
function getPage(id) {
  return (dispatch) => {
    dispatch(getPageStart(id));

    return admin.getPage(id).then((data) => {
      dispatch(getPageSuccess(id, data));
    }, (error) => {
      dispatch(getPageFailure(id, error));
    });
  };
}

const getChildrenStart = createAction('GET_CHILDREN_START', id => ({ id }));
const getChildrenSuccess = createAction('GET_CHILDREN_SUCCESS', (id, items, meta) => ({ id, items, meta }));
const getChildrenFailure = createAction('GET_CHILDREN_FAILURE', (id, error) => ({ id, error }));

/**
 * Gets the children of a node from the API.
 */
function getChildren(id, offset = 0) {
  return (dispatch) => {
    dispatch(getChildrenStart(id));

    return admin.getPageChildren(id, {
      offset: offset,
    }).then(({ items, meta }) => {
      const nbPages = offset + items.length;
      dispatch(getChildrenSuccess(id, items, meta));

      // Load more pages if necessary. Only one request is created even though
      // more might be needed, thus naturally throttling the loading.
      if (nbPages < meta.total_count && nbPages < MAX_EXPLORER_PAGES) {
        dispatch(getChildren(id, nbPages));
      }
    }, (error) => {
      dispatch(getChildrenFailure(id, error));
    });
  };
}

const openExplorer = createAction('OPEN_EXPLORER', id => ({ id }));
export const closeExplorer = createAction('CLOSE_EXPLORER');

export function toggleExplorer(id) {
  return (dispatch, getState) => {
    const { explorer, nodes } = getState();

    if (explorer.isVisible) {
      dispatch(closeExplorer());
    } else {
      const page = nodes[id];

      dispatch(openExplorer(id));

      if (!page) {
        dispatch(getChildren(id));
      }

      // We need to get the title of the starting page, only if it is not the site's root.
      const isNotRoot = id !== 1;
      if (isNotRoot) {
        dispatch(getPage(id));
      }
    }
  };
}

export const popPage = createAction('POP_PAGE');
const pushPagePrivate = createAction('PUSH_PAGE', id => ({ id }));

export function pushPage(id) {
  return (dispatch, getState) => {
    const { nodes } = getState();
    const page = nodes[id];

    dispatch(pushPagePrivate(id));

    if (page && !page.isFetching && !page.children.count > 0) {
      dispatch(getChildren(id));
    }
  };
}
