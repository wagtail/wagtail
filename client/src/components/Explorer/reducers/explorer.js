const defaultState = {
  isVisible: false,
  path: [],
};

/**
 * Oversees the state of the explorer. Defines:
 * - Where in the page tree the explorer is at.
 * - Whether the explorer is open or not.
 */
export default function explorer(prevState = defaultState, { type, payload }) {
  const state = Object.assign({}, prevState);

  switch (type) {
  case 'OPEN_EXPLORER':
    // Provide a starting page when opening the explorer.
    state.path = [payload.id];
    state.isVisible = true;
    break;

  case 'CLOSE_EXPLORER':
    state.path = [];
    state.isVisible = false;
    break;

  case 'PUSH_PAGE':
    state.path = state.path.concat([payload.id]);
    break;

  case 'POP_PAGE':
    state.path = state.path.slice(0, -1);
    break;

  default:
    break;
  }

  return state;
}
