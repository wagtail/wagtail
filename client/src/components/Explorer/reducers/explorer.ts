export interface State {
  isVisible: boolean;
  path: number[];
}

const defaultState: State = {
  isVisible: false,
  path: [],
};

export const OPEN_EXPLORER = 'OPEN_EXPLORER';
interface OpenExplorerAction {
  type: typeof OPEN_EXPLORER;
  payload: {
    id: number;
  }
}

export const CLOSE_EXPLORER = 'CLOSE_EXPLORER';
interface CloseExplorerAction {
  type: typeof CLOSE_EXPLORER;
}

export const PUSH_PAGE = 'PUSH_PAGE';
interface PushPageAction {
  type: typeof PUSH_PAGE;
  payload: {
    id: number;
  }
}

export const POP_PAGE = 'POP_PAGE';
interface PopPageAction {
  type: typeof POP_PAGE;
}

export type Action = OpenExplorerAction | CloseExplorerAction | PushPageAction | PopPageAction;

/**
 * Oversees the state of the explorer. Defines:
 * - Where in the page tree the explorer is at.
 * - Whether the explorer is open or not.
 */
export default function explorer(prevState = defaultState, action: Action) {
  switch (action.type) {
  case OPEN_EXPLORER:
    // Provide a starting page when opening the explorer.
    return {
      isVisible: true,
      path: [action.payload.id],
    };

  case CLOSE_EXPLORER:
    return defaultState;

  case PUSH_PAGE:
    return {
      isVisible: prevState.isVisible,
      path: prevState.path.concat([action.payload.id]),
    };

  case POP_PAGE:
    return {
      isVisible: prevState.isVisible,
      path: prevState.path.slice(0, -1),
    };

  default:
    return prevState;
  }
}
