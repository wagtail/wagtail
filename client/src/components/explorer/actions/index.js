import { createAction } from 'redux-actions';

import { API_PAGES, PAGES_ROOT_ID } from 'config';

function _getHeaders() {
  const headers = new Headers();
  headers.append('Content-Type', 'application/json');

  return {
    credentials: 'same-origin',
    headers: headers,
    method: 'GET'
  };
}

function _get(url) {
  return fetch(url, _getHeaders()).then(response => response.json());
}

export const fetchStart = createAction('FETCH_START');

export const fetchSuccess = createAction('FETCH_SUCCESS', (id, body) => {
  return { id, body };
});

export const fetchFailure = createAction('FETCH_FAILURE');

export const pushPage = createAction('PUSH_PAGE');

export const popPage = createAction('POP_PAGE');

export const fetchBranchSuccess = createAction('FETCH_BRANCH_SUCCESS', (id, json) => {
  return { id, json };
});

export const fetchBranchStart = createAction('FETCH_BRANCH_START');

export const clearError = createAction('CLEAR_TRANSPORT_ERROR');

export const resetTree = createAction('RESET_TREE');

export const treeResolved = createAction('TREE_RESOLVED');

// Make this a bit better... hmm....
export function fetchTree(id = 1) {
  return (dispatch) => {
    dispatch(fetchBranchStart(id));

    return _get(`${API_PAGES}${id}/`)
      .then(json => {
        dispatch(fetchBranchSuccess(id, json));

        // Recursively walk up the tree to the root, to figure out how deep
        // in the tree we are.
        if (json.meta.parent) {
          dispatch(fetchTree(json.meta.parent.id));
        } else {
          dispatch(treeResolved());
        }
      });
  };
}

export function fetchRoot() {
  return (dispatch) => {
    // TODO Should not need an id.
    dispatch(resetTree(1));

    return _get(`${API_PAGES}?child_of=${PAGES_ROOT_ID}`)
      .then(json => {
        // TODO right now, only works for a single homepage.
        // TODO What do we do if there is no homepage?
        const rootId = json.items[0].id;

        dispatch(fetchTree(rootId));
      });
  };
}

export const toggleExplorer = createAction('TOGGLE_EXPLORER');

export const fetchChildrenSuccess = createAction('FETCH_CHILDREN_SUCCESS', (id, json) => {
  return { id, json };
});

export const fetchChildrenStart = createAction('FETCH_CHILDREN_START');

/**
 * Gets the children of a node from the API
 */
export function fetchChildren(id = 'root') {
  return (dispatch, getState) => {
    const { explorer } = getState();

    let api = `${API_PAGES}?child_of=${id}`;

    if (explorer.fields) {
      api += `&fields=${explorer.fields.map(global.encodeURIComponent).join(',')}`;
    }

    if (explorer.filter) {
      api = `${api}&${explorer.filter}`;
    }

    dispatch(fetchChildrenStart(id));

    return _get(api)
      .then(json => dispatch(fetchChildrenSuccess(id, json)));
  };
}

export function setFilter(filter) {
  return (dispatch, getState) => {
    const { explorer } = getState();
    const id = explorer.path[explorer.path.length - 1];

    dispatch({
      payload: {
        filter,
        id
      },
      type: 'SET_FILTER'
    });

    dispatch(fetchChildren(id));
  };
}

/**
 * TODO: determine if page is already loaded, don't load it again, just push.
 */
export function fetchPage(id = 1) {
  return dispatch => {
    dispatch(fetchStart(id));
    return _get(`${API_PAGES}${id}/`)
      .then(json => dispatch(fetchSuccess(id, json)))
      .then(json => dispatch(fetchChildren(id, json)))
      .catch(json => dispatch(fetchFailure(new Error(JSON.stringify(json)))));
  };
}

export const setDefaultPage = createAction('SET_DEFAULT_PAGE');
