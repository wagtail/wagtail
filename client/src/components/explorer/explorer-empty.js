import React from 'react';


const ExplorerEmpty = () => {
  return <div className='c-explorer__placeholder'>No results</div>
}

export default ExplorerEmpty;
