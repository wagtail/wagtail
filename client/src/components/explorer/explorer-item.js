import React, { Component, PropTypes } from 'react';
import StateIndicator from 'components/state-indicator';

export default class ExplorerItem extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    const { title, data } = this.props;

    return (
      <div className="c-explorer__item">
        <h3 className="c-explorer__title">
          <StateIndicator state={data.state} />
          {title}
        </h3>
      </div>
    );
  }
}


ExplorerItem.propTypes = {
  title: PropTypes.string,
  data: PropTypes.object
};
