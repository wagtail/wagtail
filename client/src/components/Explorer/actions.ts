import { ThunkAction } from 'redux-thunk';

import * as admin from '../../api/admin';
import { createAction } from '../../utils/actions';
import { MAX_EXPLORER_PAGES } from '../../config/wagtailConfig';

import { State as ExplorerState, Action as ExplorerAction } from './reducers/explorer';
import { State as NodeState, WagtailPageAPI, Action as NodeAction } from './reducers/nodes';

interface State {
  explorer: ExplorerState,
  nodes: NodeState,
}

type Action = ExplorerAction | NodeAction;
type ThunkActionType = ThunkAction<void, State, unknown, Action>;

const getPageSuccess = createAction('GET_PAGE_SUCCESS', (id: number, data: WagtailPageAPI) => ({ id, data }));
const getPageFailure = createAction('GET_PAGE_FAILURE', (id: number, error: Error) => ({ id, error }));

/**
 * Gets a page from the API.
 */
function getPage(id: number): ThunkActionType {
  return (dispatch) => admin.getPage(id).then((data) => {
    dispatch(getPageSuccess(id, data));
  }, (error) => {
    dispatch(getPageFailure(id, error));
  });
}

const getChildrenStart = createAction('GET_CHILDREN_START', (id: number) => ({ id }));
const getChildrenSuccess = createAction(
  'GET_CHILDREN_SUCCESS',
  (id, items: WagtailPageAPI[], meta: any) => ({ id, items, meta })
);
const getChildrenFailure = createAction('GET_CHILDREN_FAILURE', (id: number, error: Error) => ({ id, error }));

/**
 * Gets the children of a node from the API.
 */
function getChildren(id: number, offset = 0): ThunkActionType {
  return (dispatch) => {
    dispatch(getChildrenStart(id));

    return admin.getPageChildren(id, {
      offset: offset,
    }).then(({ items, meta }) => {
      const nbPages = offset + items.length;
      dispatch(getChildrenSuccess(id, items, meta));

      // Load more pages if necessary. Only one request is created even though
      // more might be needed, thus naturally throttling the loading.
      if (nbPages < meta.total_count && nbPages < MAX_EXPLORER_PAGES) {
        dispatch(getChildren(id, nbPages));
      }
    }, (error) => {
      dispatch(getChildrenFailure(id, error));
    });
  };
}

const openExplorer = createAction('OPEN_EXPLORER', id => ({ id }));
export const closeExplorer = createAction('CLOSE_EXPLORER');

export function toggleExplorer(id: number): ThunkActionType {
  return (dispatch, getState) => {
    const { explorer, nodes } = getState();

    if (explorer.isVisible) {
      dispatch(closeExplorer());
    } else {
      const page = nodes[id];

      dispatch(openExplorer(id));

      if (!page) {
        dispatch(getChildren(id));
      }

      // We need to get the title of the starting page, only if it is not the site's root.
      const isNotRoot = id !== 1;
      if (isNotRoot) {
        dispatch(getPage(id));
      }
    }
  };
}

export const popPage = createAction('POP_PAGE');
const pushPagePrivate = createAction('PUSH_PAGE', (id: number) => ({ id }));

export function pushPage(id: number): ThunkActionType {
  return (dispatch, getState) => {
    const { nodes } = getState();
    const page = nodes[id];

    dispatch(pushPagePrivate(id));

    if (page && !page.isFetching && !(page.children.count > 0)) {
      dispatch(getChildren(id));
    }
  };
}
