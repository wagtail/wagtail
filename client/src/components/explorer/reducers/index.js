import { combineReducers } from 'redux';
import explorer from './explorer';
import nodes from './nodes';
import transport from './transport';


const rootReducer = combineReducers({
  explorer,
  transport,
  nodes,
});

export default rootReducer;
