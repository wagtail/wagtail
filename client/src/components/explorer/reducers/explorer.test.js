import * as actions from '../actions';
import _ from 'lodash';
import rootReducer from './index';
import explorer from './explorer';

describe('explorer', () => {
  const initialState = {
    isVisible: false,
    isFetching: false,
    isResolved: false,
    path: [],
    currentPage: 1,
    defaultPage: 1,
    fields: ['title', 'latest_revision_created_at', 'status', 'descendants', 'children'],
    pageTypes: {},
  };

  it('exists', () => {
    expect(explorer).toBeDefined();
  });
  it('returns the initial state if no input is provided', () =>  {
    expect(explorer(undefined, undefined))
      .toEqual(initialState);
  });
  it('sets the default page', () => {
    expect(explorer(initialState, {type: 'SET_DEFAULT_PAGE', payload: 100}))
      .toEqual(_.assign({}, initialState, {defaultPage: 100}))
  });
  it('resets the tree', () => {
    expect(explorer(initialState, {type: 'RESET_TREE', payload: 100}))
      .toEqual(_.assign({}, initialState, {isFetching: true, currentPage: 100}))
  });
  it('has resolved the tree', () => {
    expect(explorer(initialState, {type: 'TREE_RESOLVED'}))
      .toEqual(_.assign({}, initialState, {isResolved: true}))
  });
  it('toggles the explorer', () => {
    expect(explorer(initialState, {type: 'TOGGLE_EXPLORER', payload: 100}))
      .toEqual(
        _.assign({}, initialState, {isVisible: !initialState.isVisible, currentPage: 100})
      )
  });
  it('starts fetching', () => {
    expect(explorer(initialState, {type: 'FETCH_START'}))
      .toEqual(_.assign({}, initialState, {isFetching: true}))
  });
  it('pushes a page to the path', () => {
    expect(explorer(initialState, {type: 'PUSH_PAGE', payload: 100}))
      .toEqual(_.assign({}, initialState, {path: initialState.path.concat([100])}))
  });
  it('pops a page off the path', () => {
    expect(explorer(_.assign({}, initialState, {path: initialState.path.concat(["root", 100])}), {type: 'POP_PAGE', payload: 100}))
      .toEqual(_.assign({}, initialState, {path: initialState.path.concat(["root"])}))
  });
});