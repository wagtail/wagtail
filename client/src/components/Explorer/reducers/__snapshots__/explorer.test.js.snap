// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`explorer OPEN_EXPLORER 1`] = `
Object {
  "isVisible": true,
  "path": Array [
    1,
  ],
}
`;

exports[`explorer POP_PAGE 1`] = `
Object {
  "isVisible": false,
  "path": Array [],
}
`;

exports[`explorer PUSH_PAGE 1`] = `
Object {
  "isVisible": false,
  "path": Array [
    100,
  ],
}
`;
